pace-util/pace/util/checkpointer/validation.py

```diff
@@ -0,0 +1,129 @@
+import collections
+import contextlib
+import os.path
+from typing import MutableMapping, Tuple
+
+import numpy as np
+
+from pace.util._optional_imports import xarray as xr
+
+from .base import Checkpointer
+from .thresholds import ArrayLike, SavepointName, SavepointThresholds, cast_to_ndarray
+
+
+def _clip_pace_array_to_target(
+    array: np.ndarray, target_shape: Tuple[int, ...]
+) -> np.ndarray:
+    """
+    Clip an array from pace to align it to a target shape from target serialized data.
+
+    Assumes the target shape has the same number of halo points
+    at the start and end of each axis.
+
+    Assumes the input array has the same number of halo points at the start and end of
+    each axis, but with an additional buffer point at the end of each axis if the
+    data is defined on cell centers.
+
+    Args:
+        array: array to clip
+        target_shape: shape of target array
+    """
+    array = _remove_buffer_if_needed(array, target_shape)
+    return _remove_symmetric_halos(array, target_shape)
+
+
+def _remove_buffer_if_needed(array: np.ndarray, target_shape: Tuple[int, ...]):
+    selection = []
+    for array_len, target_len in zip(array.shape, target_shape):
+        if (array_len - target_len) % 2 == 1:
+            # clip the buffer point
+            selection.append(slice(0, -1))
+        else:
+            selection.append(slice(None, None))
+    return array[tuple(selection)]
+
+
+def _remove_symmetric_halos(array: np.ndarray, target_shape: Tuple[int, ...]):
```

Comment:
Is there a reason we don't use the number of halos to determine how to clip?

Reply:
Because we don't have access to this information. Not every "pace array" is a quantity, and the fortran data can have anything between 0 and 3 halo points.